quota/mysql/mysql_quota.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package mysql defines a MySQL-based quota.Manager implementation.
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+
+	"github.com/google/trillian/quota"
+)
+
+const (
+	// DefaultMaxUnsequenced is a suggested value for MaxUnsequencedRows.
+	// Note that this is a Global/Write quota suggestion, so it applies across trees.
+	DefaultMaxUnsequenced = 500000 // About 2h of non-stop signing at 70QPS.
+
+	countFromInformationSchemaQuery = `
+		SELECT table_rows
+		FROM information_schema.tables
+		WHERE table_schema = schema()
+			AND table_name = ?
+			AND table_type = ?`
+	countFromUnsequencedQuery = "SELECT COUNT(*) FROM Unsequenced"
+)
+
+// QuotaManager is a MySQL-based quota.Manager implementation.
+//
+// It has two working modes: one queries the information schema for the number of Unsequenced rows,
+// the other does a select count(*) on the Unsequenced table. Information schema queries are
+// default, even though they are approximate, as they're constant time (select count(*) on InnoDB
+// based MySQL needs to traverse the index and may take quite a while to complete).
+//
+// QuotaManager only implements Global/Write quotas, which is based on the number of Unsequenced
+// rows (to be exact, tokens = MaxUnsequencedRows - actualUnsequencedRows).
+// Other quotas are considered infinite.
+type QuotaManager struct {
+	DB                 *sql.DB
+	MaxUnsequencedRows int
+	UseSelectCount     bool
+}
+
+// GetUser implements quota.Manager.GetUser.
+// User quotas are not implemented by QuotaManager.
+func (m *QuotaManager) GetUser(ctx context.Context, req interface{}) string {
+	return "" // Not used
+}
+
+// GetTokens implements quota.Manager.GetTokens.
+// It doesn't actually reserve or retrieve tokens, instead it allows access based on the number of
+// rows in the Unsequenced table.
+func (m *QuotaManager) GetTokens(ctx context.Context, numTokens int, specs []quota.Spec) error {
+	for _, spec := range specs {
+		if spec.Group != quota.Global || spec.Kind != quota.Write {
+			continue
+		}
+		// Only allow global writes if Unsequenced is under the expected limit
+		count, err := m.countUnsequenced(ctx)
+		if err != nil {
+			return err
+		}
+		if count+numTokens > m.MaxUnsequencedRows {
```

Comment:
If it's not for production use it's not as important. It might not be a few requests over the limit e.g. if 25 people ask for 1000 at the same time. Might be worth bounding how many tokens can be taken at once.